Add first release test - Please DO NOT MERGE

newtewt opened the pull request:
Please DO NOT MERGE

# Description

Trying to add the first of a few release tests to try and automate that process. Looking for a review and suggestions. I'll list my concerns and hope for any other review. 

1. I'm concerned about using base.conf.js as it always set the user as admin and thinking about pulling out what I can to be reusable by the conf.js. I always can then remove the additional code that applies a config. Wanted to discuss this before moving forward. 

1. I think we should look into reworking the conf.js so that we use suites instead of a separate conf.js. The only differenence I can see is that we pass in a param for the source dir and that can easily be done using suites. 

1. I am getting the standard config applied however it seems like it is not waiting long enough before executing the tests that the config is fully applied. I currently have a poor hack in place(tests/apply-config.js) to try and slow the async but need help on this one. We'll need to be able to apply other configs as we go for different tests.  

1. For now please ignore the test as it does very little. I just wanted to tackle getting a test running against the config for now. 

 

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

garethbowen commented on tests/apply-config.js:
```diff
@@ -0,0 +1,55 @@
+
+
+
+module.exports = applyConfig;
```
I think we should refactor medic-conf so a function is exposed which does this. It's essentially the `lib/main.js` except without the CLI parsing stuff. I think that'd be useful for other scripts as well so it's a good option to have.

garethbowen replied:
Yes precisely.

garethbowen commented on tests/base.conf.js:
```diff
@@ -50,10 +51,17 @@ class BaseConfig {
         browser.driver.wait(startApi(), 135 * 1000, 'API took too long to start up');
-        browser.driver.sleep(10000); // wait for startup to complete
+        browser.driver.sleep(10000).then(() => {
+          if (pathToConfig){
```
I don't think this needs to be in `then` - I think protractor will automatically pause execution until the sleep finishes.

garethbowen commented on tests/page-objects/common/common.po.js:
```diff
@@ -92,10 +92,10 @@ module.exports = {
+  goToPeople: async () => {
+    await browser.get(utils.getBaseUrl() + 'contacts/');
+    await helper.waitUntilReady(medicLogo);
+    await helper.waitUntilReady(element(by.id('contacts-list')));
```
These awaits are implicit in protractor - you shouldn't need them.

garethbowen commented on tests/page-objects/contacts/contacts.po.js:
```diff
@@ -27,7 +29,7 @@ module.exports = {
     writeName.click();
     newPlaceName.sendKeys(districtName);
-    submitButton.click();
+    await submitButton.click();
```
Same here. We don't use them anywhere else - why is it necessary here?

garethbowen commented on tests/release.tests.conf.js:
```diff
@@ -0,0 +1,7 @@
+const BaseConfig = require('./base.conf.js');
+console.log(__dirname);
```
Accidentally left in?

garethbowen commented on tests/release.tests.conf.js:
```diff
@@ -0,0 +1,7 @@
+console.log(__dirname);
+const pathToStandardConfig = __dirname + '/../config/standard';
+let baseConfig = new BaseConfig('e2e/release', { headless:false }, pathToStandardConfig);
+baseConfig.SELENIUM_PROMISE_MANAGER = false;
```
Oh.. this is why you need async everywhere I guess. What's the advantage of turning this off?

garethbowen replied:
Ok, yes, that's much better. Nice one.

garethbowen commented on tests/release.tests.conf.js:
```diff
@@ -0,0 +1,7 @@
+const BaseConfig = require('./base.conf.js');
+console.log(__dirname);
+const pathToStandardConfig = __dirname + '/../config/standard';
+let baseConfig = new BaseConfig('e2e/release', { headless:false }, pathToStandardConfig);
```
I'd use headless mode when you're ready to merge.

garethbowen commented on Gruntfile.js:
```diff
@@ -999,6 +1004,11 @@ module.exports = function(grunt) {
+  grunt.registerTask('release-e2e', 'Deploy app for testing and run e2e tests', [
+    'e2e-deploy',
+    'protractor:release-e2e-tests',
+  ]);
```
This should be executed as part of CI for every build.

garethbowen commented on Gruntfile.js:
```diff
@@ -739,6 +739,11 @@ module.exports = function(grunt) {
     protractor: {
+      'release-e2e-tests':{
+        options:{
+          configFile: 'tests/relase.tests.conf.js'
```
There's a typo in the config file name.

Also, I think it should be called `regression-tests.conf.js` because I want to run it every build, not just pre-release. If it ends up taking way too long then we can change it up but for the near future we should run it as early as possible.

Thirdly, I don't think this should be a separate suite to the other e2e test suite - they're all regression tests. Right now it sort of has to be because we're using different config, but are you envisioning one day all tests will run against the standard config?

garethbowen commented:
> why do we have a grunt task for debugging and not using IDE to do it?

I think the grunt task only changes the headlessness and doesn't actually "debug". How do you use the IDE?

garethbowen commented:
Sounds good. Not everyone uses an IDE so I think we should support both. It's probably a badly named grunt task and should be `e2e-headless` or something to avoid confusion.

newtewt commented:
@garethbowen  can you review the latest changes? 

Removed the release specific conf settings and made everything apart of e2e general release. 

Removed applying config as of now because CHT will be the default soon enough and I can revisit to update the current test. 

Updated medic-conf to allow requiring of the executeAction in code which allows us to apply the configs. Applyconfig is now way reduced. 

Now that we have a place to add tests I'll continue with adding the admin user specific tests. Also, my thought is we are going to have to convert our release tests for CHT so might as well update the github tests and then write a automated test for them.
